torch/fx/experimental/symbolic_shapes.py

```diff
@@ -2998,19 +3066,37 @@ def hint(s):
                 sexpr = ShapeGuardPrinter(symbol_to_source, source_ref, self.var_to_sources).doprint(expr)
                 exprs.append(f"{source_ref(source)} == {sexpr}")
                 if (
-                    isinstance(expr, sympy.Symbol) and
-                    expr in symbol_to_constraints and
                     isinstance(source, TensorPropertySource)
                     and source.prop is TensorProperty.SIZE
-                    and equalities_inputs and
-                    not equalities_inputs.is_equal(source, symbol_to_source[expr][0])
+                    and equalities_inputs
+                    and len(expr.free_symbols) == 1
                 ):
-                    msg = (
-                        f"The values of {self._debug_name(source)} = {source.name()} and "
-                        f"{self._debug_name(symbol_to_source[expr][0])} = {symbol_to_source[expr][0].name()} "
-                        "must always be equal."
-                    )
-                    record_constraint_violation(equalities_inputs.warn_only, self._debug_name(source), msg)
+                    symbol = next(iter(expr.free_symbols))
+                    if (
+                        isinstance(expr, sympy.Symbol) and
+                        expr in symbol_to_constraints and
+                        not equalities_inputs.is_equal(source, symbol_to_source[expr][0])
+                    ):
+                        msg = (
+                            f"The values of {self._debug_name(source)} = {source.name()} and "
+                            f"{self._debug_name(symbol_to_source[expr][0])} = {symbol_to_source[expr][0].name()} "
+                            "must always be equal."
+                        )
+                        record_constraint_violation(equalities_inputs.warn_only, self._debug_name(source), msg)
+
+                    if (
+                        not isinstance(expr, sympy.Symbol) and
+                        symbol in symbol_to_constraints and
+                        not equalities_inputs.is_derived(source, symbol, expr)
+                    ):
+                        src = symbol_to_source[symbol][0]
+                        msg = (
+                            f"The values of {self._debug_name(source)} = {source.name()} must always be related to "
+                            f"the values of {self._debug_name(src)} = {src.name()} by "
+                            f"{self._debug_name(source)} = {expr.subs({symbol: sympy.Symbol(self._debug_name(src))})}."
+                        )
+                        record_constraint_violation(equalities_inputs.warn_only, self._debug_name(source), msg)
```

Comment:
I think I'm just being lazy here. I should implement some transitive chaining of linear constraints. This *should not* be difficult.

Reply:
Actually I might have a proof sketch when the root is a phantom symbol because we do some normalization already; we transitively propagate all derived dims based off such a symbol.

But for something that depends on an input source, here's a repro that highlights the lack of transitive equality reasoning you mention:

```
    class Foo(torch.nn.Module):
        def forward(self, x, y, z):
            return x + y[1:] + z[2:]

    foo = Foo()

    u, v, w = torch.randn(5), torch.randn(6), torch.randn(7)
    ep = torch.export.export(
        foo,
        (u, v, w),
        dynamic_shapes=({0: dimx}, {0: dimy}, {0: dimz}),
        strict=strict,
    )
    print(ep)
```

Dumb error follows!

```
torch._dynamo.exc.UserError: Constraints violated ((dimy+1), dimy)! For more information, run with TORCH_LOGS="+dynamic".
  - The values of dimy = L['y'].size()[0] must always be related to the values of (dimy-1) = L['x'].size()[0] by dimy = (dimy-1) + 1.
  - The values of (dimy+1) = L['z'].size()[0] must always be related to the values of (dimy-1) = L['x'].size()[0] by (dimy+1) = (dimy-1) + 2.

Suggested fixes:
  (dimy+1) = (dimy-1) + 2
  dimy = (dimy-1) + 1
```

After I fix this I will try to articulate why this can't happen with phantom symbols or come up with another repro for that and fix it.

Reply:
This should now be fixed.